action.yml

```diff
@@ -220,7 +221,7 @@ runs:
           ${{ inputs.stack-arguments }}
 
     - name: Save dependencies cache
-      if: steps.restore-deps.outputs.cache-hit != 'true'
+      if: success() || (failure() && inputs.cache-save-always)
```

Comment:
[`success()` is applied by default][0] when an `if` conditional does not use a status check function. [`failure()` is needed][1] in order to run the step after some previous failure, which removes the default `success()` and requires it to be explicit.

Using `outputs.cache-hit` with the status check functions would be:

```suggestion
      if: (success() && steps.restore-deps.outputs.cache-hit != 'true') || (failure() && inputs.cache-save-always)
```

That affects saving the cache when a cache hit occurred and the build succeeded (no longer no-ops). If `inputs.cache-save-always` is used by the action, then it's possible the cache is a partial set of artifacts from a previous failed build. In this case, the successful build has added artifacts that will remain uncached.

[0]: https://docs.github.com/en/actions/learn-github-actions/expressions#status-check-functions
[1]: https://docs.github.com/en/actions/learn-github-actions/expressions#failure-with-conditions

Reply:
> when an if conditional does not use a status check function

I didn't know that part. That makes it complicated, but I guess I get it.

> That affects saving the cache when a cache hit occurred and the build succeeded (no longer no-ops). If inputs.cache-save-always is used by the action, then it's possible the cache is a partial set of artifacts from a previous failed build. In this case, the successful build has added artifacts that will remain uncached

I'm having trouble parsing this. You're saying the default behavior (using this action with `cache-save-always: false`) is changed or not? If it is changed, is there a reason it can't be not-changed?

Reply:
The default behavior with `cache-save-always: false` is not changed, but using `outputs.cache-hit` as suggested introduces the aforementioned problem when `cache-save-always: true`. One way to address it would be to use a different `key` when saving the cache after a build failure, versus what is currently used when the build has succeeded, and using this different `key` in `restore-keys`. This would allow restoring the use of `outputs.cache-hit` in both the build and the cache-save steps, but I don't think the added complexity is worthwhile. `stack build` should _essentially_ noop if a complete cache of build artifacts was previously restored.

Reply:
> The default behavior with `cache-save-always: false` is not changed

Great, love it. Merge away and minor version bump :hear_no_evil: 

> but

I think what you're saying here is just that `cache-save-always: true` results in us saving the cache even on a primary key hit. Right? If so, great, love it, don't even consider it an issue. The option literally says to save "always", which sounds like what that is.

Reply:
> I think what you're saying here is just that `cache-save-always: true` results in us saving the cache even on a primary key hit. Right?

Not exactly. The semantics of the input is to save the cache _even when the build has failed_, regardless of the primary key hit.

I have some confusion about what is considered "behavior" and therefore what changes to the default behavior necessitate a major version bump. If this PR was released as a minor change, workflows using `@v4` would see exactly four changes when inspecting the logs:

- When the dependencies cache is restored via the primary key, the step to build dependencies runs.
- When the dependencies cache is restored via the primary key, the step to save the artifacts from building dependencies runs.
- When the build cache is restored via the primary key, the step to build the project runs.
- When the build cache is restored via the primary key, the step to save the artifacts from building the project runs.

Previously, each of the steps in the aforementioned scenario would no-op. I _think_ that this is not a meaningful difference in behavior, because the expectation is that the action builds/tests a project. Running `stack build` or saving the cache, instead of a no-op, should have no further impact than the short interval where `stack build` determines there is nothing left to build.

The argument from the other direction and the reason I initially recommend a major version bump, is that going from a no-op to a step that runs _anything_ is a place where bugs could now arise, however unlikely that might be.

My opinion is that a workflow using something other than a fully-specified version or hash opts-in to some possibility that upstream updates result in some bugs, and trust that maintainers try not to introduce those bugs. In this case, a major version bump seems overly defensive.

Reply:
I think I see.

"behavior" here, at least the behavior I care about, are steps that were a no-op before, and no do things. That's a pretty important behavior. Because you've listed 4 steps that would previously no-op and now do things, then when you said this:

> The default behavior with `cache-save-always: false` is not changed

I think that's NOT true.

> The argument from the other direction and the reason I initially recommend a major version bump, is that going from a no-op to a step that runs anything ~is a place where bugs could now arise, however unlikely that might be.~
>
> My opinion is that a workflow using something other than a fully-specified version or hash opts-in to some possibility that upstream updates result in some bugs

I agree, but this is not related to bugs, it's because steps take time. We work pretty hard on this action to make it not do unnecessary work and therefore be as fast as possible. Being able to opt into a configuration where you cache more, but possibly run slower is great, but having the default behavior change such that more work is done and therefore the action takes longer is not.

And this, IMO, is not about major-vs-minor, it's actually about, can we please try to avoid that? I'm not sure why we can't keep the no-op behavior when `cache-save-always: false` we have all the information there, why can't we build the appropriate `if` to get there?

Reply:
> can we please try to avoid that? I'm not sure why we can't keep the no-op behavior when `cache-save-always: false`; we have all the information there, why can't we build the appropriate `if` to get there?

Sure. Restoring the no-op behavior requires being able to differentiate the cache saved after a build failure versus after a success. I've added a `-partial` suffix to be set on the cache key to achieve this, but this requires a little added complexity in the form of an additional step to set the suffix, since status check functions can only be used in an `if` conditional. These steps are only run when `cache-save-always: true`, so nothing changes in the default behavior.

The workflow in this repo shows the action run without this new input, with or without a full cache-hit, and the subsequent keys used when this suffix is not set. https://github.com/freckle/megarepo/pull/32424 has a CI run with [`cache-save-always: false` and build failure][0] (no cache is saved) and [`cache-save-always: true` and build failure][1] (saving with a `-partial` key).

[0]: https://github.com/freckle/megarepo/actions/runs/7616139302/job/20742234779?pr=32424
[1]: https://github.com/freckle/megarepo/actions/runs/7616139302/job/20742235580

Reply:
After some conversation over Slack, I've reverted some changes around using different cache keys and have instead extended the `if` conditionals toward some new behavior when the added input is set, which should keep the default behavior without this input the same.